### PR TITLE
feat: Add support for Unix sockets.

### LINKED ISSUE #47: Add support for unix sockets
implement: spec.instances.socketType=UNIX, unixSocketDirectory

implement: spec.instances.unixSocketPathEnvName

### PR DESCRIPTION
**Pull request by hessjcg**
The operator will configure the pod with a proxy that listens on a unix socket. The operator will mount the
parent directory of each instance's `unixSocketPath` to each container in the pod. When the proxy starts, it 
will create a unix socket for the instance in that mounted directory.

Care should be take to ensure that the parent directory can safely be used across all containers in the pod.
 
Fixes #47

### CODE REVIEW

File: config/crd/bases/cloudsql.cloud.google.com_authproxyworkloads.yaml [hunk 1]
```diff
@@ -927,6 +927,12 @@ spec:
                       privateIP:
                         description: PrivateIP Enable connection to the Cloud SQL instance's private ip for this instance. Optional, default false.
                         type: boolean
+                      unixSocketPath:
```
Comment (hessjcg) [inline config/crd/bases/cloudsql.cloud.google.com_authproxyworkloads.yaml:930]: Generated from source comments in  authproxyworkload_types.go

File: installer/cloud-sql-proxy-operator.yaml [hunk 2]
```diff
@@ -945,6 +945,12 @@ spec:
                       privateIP:
                         description: PrivateIP Enable connection to the Cloud SQL instance's private ip for this instance. Optional, default false.
                         type: boolean
+                      unixSocketPath:
```
Comment (hessjcg) [inline installer/cloud-sql-proxy-operator.yaml:948]: Generated from source comments in  authproxyworkload_types.go

**Review by hessjcg — COMMENTED** [review]
This can't be merged until after the Feb 2023 proxy release, which adds support for the unix-socket-path instance parameter.

File: Makefile [hunk 3]
```diff
@@ -375,6 +375,23 @@ e2e_image_push: generate # Build and push a operator image to the e2e artifact r
...
 	$(PWD)/tools/docker-build.sh
 
+###
+# Build a version of the cloud-sql-proxy from local sources
+E2E_LOCAL_PROXY_PROJECT_DIR?=/not-set
+E2E_LOCAL_PROXY_BUILD_URL_FILE=$(PWD)/bin/last-local-proxy-url.txt
+E2E_LOCAL_PROXY_BUILD_URL=$(shell cat $(E2E_LOCAL_PROXY_BUILD_URL_FILE) | tr -d "\n")
+
+.PHONY: e2e_image_push
+e2e_local_proxy_image_push: # Build and push the proxy image from a local working directory to the e2e artifact repo
```
Comment (enocom) [inline Makefile:385]: Why this addition?

File: build.sample.env [hunk 4]
```diff
@@ -6,3 +6,9 @@
 ##
 # The gcloud project id to use for the end-to-end tests
 E2E_PROJECT_ID = my-gcloud-project
+
+##
+# To run the e2e tests using the proxy code from your local machine,
+# Uncomment this line and set the path to your local working directory
+# of the cloud-sql-proxy project.
+# E2E_LOCAL_PROXY_PROJECT_DIR = /home/me/projects/cloud-sql-proxy
```
Comment (enocom) [inline build.sample.env:14]: This seems like a separate PR.

File: internal/testhelpers/resources.go [hunk 5]
```diff
@@ -89,6 +89,20 @@ func BuildPgPodSpec(mainPodSleep int, appLabel, secretName string) corev1.PodTem
 	return buildConnectPodSpec(mainPodSleep, appLabel, secretName, livenessCmd, passEnvVarName, imageName)
 }
 
+// BuildPgUnixPodSpec creates a podspec specific to Postgres databases that will
+// connect via a unix socket and run a trivial. It also configures the pod's
```
Comment (enocom) [inline internal/testhelpers/resources.go:93]: Incomplete sentence here.

File: internal/testhelpers/resources.go [hunk 6]
```diff
@@ -572,8 +600,38 @@ func (cc *TestCaseClient) CreateDeploymentReplicaSetAndPods(ctx context.Context,
 }
 
 // BuildAuthProxyWorkload creates an AuthProxyWorkload object with a
-// single connection instance.
+// single instance with a tcp connection.
 func BuildAuthProxyWorkload(key types.NamespacedName, connectionString string) *v1alpha1.AuthProxyWorkload {
+	p := NewAuthProxyWorkload(key)
+	AddTCPInstance(p, connectionString)
```
Comment (enocom) [inline internal/testhelpers/resources.go:606]: Why return something if we're not going to use it?

File: internal/api/v1alpha1/authproxyworkload_types.go [hunk 7]
```diff
@@ -239,6 +239,16 @@ type InstanceSpec struct {
 	// Optional, when set this environment variable will be added to all containers in the workload.
 	//+kubebuilder:validation:Optional
 	HostEnvName string `json:"hostEnvName,omitempty"`
+
+	// UnixSocketPath is the path to the unix socket where the proxy will listen
+	// for connnections. This will be mounted to all containers in the pod.
+	//+kubebuilder:validation:Optional
+	UnixSocketPath string `json:"unixSocketPath,omitempty"`
```
Comment (enocom) [inline internal/api/v1alpha1/authproxyworkload_types.go:246]: Why do we have this field and the one below?

File: internal/workload/podspec_updates.go [hunk 8]
```diff
@@ -475,20 +478,44 @@ func (s *updateState) updateContainer(p *cloudsqlapi.AuthProxyWorkload, wl Workl
 		params := map[string]string{}
 
 		// if it is a TCP socket
+		if inst.UnixSocketPath == "" {
+
+			port := s.useInstancePort(p, inst)
+			params["port"] = fmt.Sprint(port)
+			if inst.HostEnvName != "" {
```
Comment (enocom) [inline internal/workload/podspec_updates.go:485]: What do we do if HostEnvName and PortEnvName are empty strings?

File: internal/workload/podspec_updates.go [hunk 9]
```diff
@@ -475,20 +478,44 @@ func (s *updateState) updateContainer(p *cloudsqlapi.AuthProxyWorkload, wl Workl
...
+			}
+			if inst.PortEnvName != "" {
+				s.addWorkloadEnvVar(p, inst, corev1.EnvVar{
+					Name:  inst.PortEnvName,
+					Value: fmt.Sprint(port),
+				})
+			}
+		} else {
+			// else if it is a unix socket
+			params["unix-socket-path"] = inst.UnixSocketPath
```
Comment (enocom) [inline internal/workload/podspec_updates.go:499]: Where are we using `params`?

File: internal/workload/podspec_updates.go [hunk 10]
```diff
@@ -613,6 +640,75 @@ func (s *updateState) addHealthCheck(_ *cloudsqlapi.AuthProxyWorkload, c *corev1
...
+	nameAccessor := func(v corev1.Volume) string {
+		return v.Name
+	}
+	thingAccessor := func(v *managedVolume) corev1.Volume {
+		return v.Volume
+	}
+	ps.Volumes = applyVolumeThings[corev1.Volume](s, ps.Volumes, nameAccessor, thingAccessor)
+}
+
+// applyVolumeThings implements complex reconcile logic that is duplicated for both
```
Comment (enocom) [inline internal/workload/podspec_updates.go:687]: Could we elaborate on what "complex" means here?

File: tests/e2e_test.go [hunk 11]
```diff
@@ -335,9 +350,20 @@ func TestPublicDBConnections(t *testing.T) {
 			wl.Deployment.Spec.Template = test.podTemplate
 			t.Log("Creating AuthProxyWorkload")
 
-			_, err = tp.CreateAuthProxyWorkload(ctx, key, appLabel, tp.ConnectionString, kind)
-			if err != nil {
-				t.Fatal(err)
+			if test.isUnixSocket {
+				p := testhelpers.NewAuthProxyWorkload(key)
```
Comment (enocom) [inline tests/e2e_test.go:354]: Should these tests also be checking that an application can reach the database? It looks like we're just checking that the pod is ready.

File: internal/workload/podspec_updates.go (hunk 10 above)
Comment (hessjcg) [reply internal/workload/podspec_updates.go]: I rewrote the comment to be more precise.

File: Makefile (hunk 3 above)
Comment (enocom) [reply Makefile:385]: Can we clean up this PR to remove what's in #208? It's a bit harder to review otherwise.

File: internal/api/v1alpha1/authproxyworkload_types.go (hunk 7 above)
Comment (enocom) [reply internal/api/v1alpha1/authproxyworkload_types.go:246]: What you said makes a lot of sense. Can we work that into the documentation here?

File: internal/workload/podspec_updates.go (hunk 8 above)
Comment (enocom) [reply internal/workload/podspec_updates.go:485]: This is super helpful. Shall we document this somewhere?

File: internal/workload/podspec_updates.go (hunk 9 above)
Comment (enocom) [reply internal/workload/podspec_updates.go:499]: Do you think we should make these parameters configurable in the environment in the proxy?

File: tests/e2e_test.go (hunk 11 above)
Comment (enocom) [reply tests/e2e_test.go:354]: Got it. I wish we could make that detail stand out some more here in the test.

File: internal/workload/podspec_updates.go (hunk 9 above)
Comment (hessjcg) [reply internal/workload/podspec_updates.go:499]: Yes, instance configuration should also be converted to EnvVars. I want to do a clean "feat: add unix sockets" PR without changing much other stuff first.

File: internal/workload/podspec_updates.go (hunk 8 above)
Comment (hessjcg) [reply internal/workload/podspec_updates.go:485]: I added a new file `docs/authproxyworkload-reference.md` which contains the comment above. In future PRs, we can
add more configuration detail about how the AuthProxyWorkload resource should be used. This can be input for
our official documentation.

File: tests/e2e_test.go (hunk 11 above)
Comment (hessjcg) [reply tests/e2e_test.go:354]: I added a comment.

File: internal/workload/podspec_updates.go (hunk 9 above)
Comment (enocom) [reply internal/workload/podspec_updates.go:499]: Right now the Proxy doesn't support configuring instance-level query params with env-vars. So we think that's valuable, we'll have to start with the proxy first.

File: tests/setup_test.go [hunk 12]
```diff
@@ -110,7 +110,7 @@ func setupTests() (func(), error) {
 	}
 
 	// Read e2e test configuration
-	proxyImageURL = loadValue("PROXY_IMAGE_URL", "../bin/last-proxy-image-url.txt", workload.DefaultProxyImage)
+	proxyImageURL = loadValue("PROXY_IMAGE_URL", "../bin/last-local-proxy-url.txt", workload.DefaultProxyImage)
```
Comment (enocom) [inline tests/setup_test.go:113]: This looks like an unrelated change.

**Review by enocom — APPROVED** [review]

File: tests/setup_test.go (hunk 12 above)
Comment (hessjcg) [reply tests/setup_test.go:113]: Good catch. I added this change to #208 where it belongs.

**Comment by hessjcg** [comment]
This PR is approved, but has to wait until the Feb 2023 release of the Proxy.
